rendermodules/rough_align/apply_rough_alignment_to_montages.py

```diff
@@ -61,39 +66,157 @@
+                results[t.tileId] = t.ip[0].maskUrl
+
+    elif mask_input_dir is not None:
+        tids = [t.tileId for t in tilespecs]
+        maskfiles = []
+        for ext in exts:
+            e = mask_input_dir + '/*.' + ext
+            maskfiles += glob.glob(e)
+        maskbasenames = np.array([os.path.basename(os.path.splitext(p)[0])
+                                  for p in maskfiles])
+        for t in tids:
+            ind = np.argwhere(maskbasenames == t).flatten()
+            if ind.size > 1:
+                raise RenderModuleException("more than one mask in "
+                                            "directory %s matches tileId %s" %
+                                            (mask_input_dir, t))
+            if ind.size == 0:
+                continue
+            results[t] = pathlib.Path(maskfiles[ind[0]]).as_uri()
+
+    return results
+
+
+def add_masks_to_lowres(render, stack, z, mask_map):
+    if len(mask_map) == 0:
+        return
+
+    tilespecs = renderapi.tilespec.get_tile_specs_from_z(
+            stack, z, render=render)
+
+    for t in tilespecs:
+        if t.tileId in mask_map:
+            t.ip['0'].maskUrl = mask_map[t.tileId]
+            renderapi.client.import_tilespecs(
+                    stack,
+                    [t],
+                    render=render)
+    return
+
+
+def filter_highres_with_masks(resolved_highres, tspec_lowres, mask_map):
+    """function to return a filtered list of tilespecs from a
+       ResolvedTiles object, based on a lowres mask
+
+    Parameters
+    ----------
+    resolved_highres : renderapi.resolvedtiles.ResolvedTiles object
+        tilespecs and transforms from a single section
+    tspec_lowres: renderapi.tilespec.TileSpec object
+        tilespec from a downsampled stack
+    mask_map: dict
+        keys should match lowres tileids, values are mask file URI
+
+    Returns
+    -------
+    new_highres: List of renderapi.tilespec.TileSpec objects
+        which highres specs are fully contained
+        within the boundary of mask=255
+    """
+
+    if tspec_lowres.tileId not in mask_map.keys():
+        return resolved_highres.tilespecs
+
+    impath = urllib.parse.unquote(
+                 urllib.parse.urlparse(
+                     mask_map[tspec_lowres.tileId]).path)
+    maskim = cv2.imread(impath, 0)
+    maskim = 255 - maskim
+    _, contours, _ = cv2.findContours(
+                         maskim,
+                         cv2.RETR_TREE,
+                         cv2.CHAIN_APPROX_SIMPLE)
+
+    mask_polygons = []
+    for c in contours:
+        c = c.squeeze().astype('float')
+        for tf in tspec_lowres.tforms:
+            c = tf.tform(c)
+        mask_polygons.append(Polygon(c).buffer(0))
+
+    new_highres = []
+    for t in resolved_highres.tilespecs:
+        tc = t.bbox_transformed(
+                    reference_tforms=resolved_highres.transforms)
+        tpoly = Polygon(tc).buffer(0)
+        for p in mask_polygons:
```

Comment:
this could be a time consuming for loop.. Have you thought about KDtree or quadtree to find the intersection of a set of polygons at a time.

Reply:
It takes about 4 seconds per section with a mask.We had 28 masks in 20,000 sections. I don't think it's worth getting fancy here